src/styles/MuiThemeProvider.js

```diff
@@ -1,20 +1,29 @@
 import {Component, PropTypes} from 'react';
 import getMuiTheme from './getMuiTheme';
+import getUniqueIdGenerator from '../utils/getUniqueIdGenerator';
+
 
 class MuiThemeProvider extends Component {
 
   static propTypes = {
     children: PropTypes.element,
     muiTheme: PropTypes.object,
+    uniqueIdGen: PropTypes.func,
   };
 
   static childContextTypes = {
     muiTheme: PropTypes.object.isRequired,
+    uniqueIdGen: PropTypes.func.isRequired,
+  };
+
+  state = {
+    uniqueIdGen: this.props.uniqueIdGen || getUniqueIdGenerator(),
```

Comment:
Yeah.😂 In fact, the feature has been discussed a bit before in https://github.com/facebook/react/issues/1137 and https://github.com/facebook/react/issues/4000. We use many hacky methods in this solution because we need a global counter which should reset itself each time React calls render method. But Material-ui is a component library and should not inject the render process.
If `MuiThemeProvider` is the root provider, many potential issues would be eliminated. But in doc it's only preferable, not necessary.